### ISSUE TITLE
Expose the query language of the underlying database.

### ISSUE THREAD

**Issue by vrde** [bigchaindb/bigchaindb#1846]
BigchainDB allows to query assets using a full text search. While this is useful for some use cases, other times devs need to exactly match some fields in an asset. Specifically, the queries you usually need to do when building an application are:
- Filter documents (and implicitly support _sorting_ and _limiting_).
- Aggregate documents.

Those operations can be done with the current _full text search_, but they require a lot of work on the client side. This wastes resources on the client side and on the server side as well (it's much cheaper to aggregate on the server and send the result to the client). _Full text search_ also has a quite neat feature: it doesn't need custom indexes to be created. There is just one big index, that is the full text search index. On the other side, allowing arbitrary queries would require us to face two big problems:
- Query language. What is the language to query BigchainDB? SQL?
- Indexes, who can create indexes? Does it _cost tokens_ to create indexes? Are there some predefined indexes developers can use without creating their own?
- Timeouts. How long can a query run? Is there a default timeout of 50ms (or whatever we think can fit the average query) and increase the timeout value if the user puts more _tokens_?

My personal opinion on the three questions:
- BigchainDB should expose the query language of the underlying database. E.g. if the backend is MongoDB, we expose the MongoDB query language.
- Ideally, a user can spend some tokens to index a specific field. Anyway there should be a set of default indexes that developers can us. We can cover basic common needs, for example:
 - `transaction.asset.data.ts`: an index on a generic timestamp value
 - `transaction.asset.data.type`: the type of the asset
 - `transaction.asset.data.tags`: an index on an array of values
 - ... and more ...
- We set a default timeout to something reasonable. If a user needs more time to run their query, they will just need to spend more tokens.

Right now, we are storing the asset in the `asset` collection. If a user spend tokens, we can create custom collections to index custom fields.

**Comment by ttmc** [comment]
When a user adds a new index, then every new write (after that) will have to update _that_ index. It slows down writes for everyone, forever.

Databases often have a limit to the number of indexes.

I like the idea of an index that everyone can use, such as `transaction.asset.data.ts`. There could even be an index on `transaction.asset.data.key1` where different users (assets) can give different meanings to key1.

**Comment by vrde** [comment]
> When a user adds a new index, then every new write (after that) will have to update that index. It slows down writes for everyone, forever.

That's true, a better approach probably would be to spend tokens to create new collections and have custom indexes there.

> I like the idea of an index that everyone can use, such as transaction.asset.data.ts. There could even be an index on transaction.asset.data.key1 where different users (assets) can give different meanings to key1.

:tada: great! About the names, I think that having something more descriptive (but still generic) helps.

**Comment by ttmc** [comment]
A similar issue is issue #1816 by @gautamdhameja

**Comment by ttmc** [comment]
An approach I shared elsewhere:  The HTTP API could provide endpoints _for the most common queries_, and leave other queries up to drivers or application developers to write using a MongoDB driver.

**Comment by ttmc** [comment]
Another option is to use a third-party package/framework to automatically generate a REST API or a GraphQL API for the database backend (MongoDB, RethinkDB, or whatever). Here's a list of some such packages/frameworks:

https://github.com/dbohdan/automatic-api

**Comment by ttmc** [comment]
In the Tendermint version of BigchainDB, each node has its own local database.

In theory, each node could allow external users to create special indexes to speed up queries (but only for the MongoDB database on that node). That would require giving external users more than just read access; they'd also require the ability to create new indexes. Maybe those could be approved manually on a case-by-case basis (since you shouldn't and can't have too many indexes).

**Comment by ttmc** [comment]
Even if nodes don't allow external users to add new indexes, it's still neat that each node can have a different set of indexes. Each node could be specialized for doing certain kinds of queries.

**Comment by gautamdhameja** [comment]
With a read-only mongo db exposed for querying, we need to explain the underlying collections (transactions, assets, metadata, blocks, votes, etc.) to the user so that they can query the specific collections. Basically, we have to document the mongo collections we are creating as part of BigchainDB init.

**Comment by vrde** [comment]
There is another big problem, that is to give tools to the client to verify that the result set belongs to the blockchain (aka: it is **valid data**), and that the result set is **complete**. This is a challenging problem.

**Comment by ttmc** [comment]
For now, we tell node operators to give users access to their node's MongoDB, maybe as a read-only user, or maybe via an API access layer that _they_ implement. It's all explained on the root docs page titled [Querying BigchainDB](https://docs.bigchaindb.com/en/latest/query.html).
